var/spack/repos/builtin/packages/zoltan/package.py

```diff
@@ -104,16 +104,23 @@ def install(self, spec, prefix):
 
             config_args.append('--with-mpi={0}'.format(spec['mpi'].prefix))
 
-            mpi_libs = self.get_mpi_libs()
-
-            # NOTE: Some external mpi installations may have empty lib
-            # directory (e.g. bg-q). In this case we need to explicitly
-            # pass empty library name.
-            if mpi_libs:
-                mpi_libs = ' -l'.join(mpi_libs)
-                config_args.append('--with-mpi-libs=-l{0}'.format(mpi_libs))
-            else:
+            # NOTE: Zoltan assumes that it's linking against an MPI library that can
+            # be found with '-lmpi,' which isn't the case for many MPI packages.
+            if os.path.basename(spec['mpi'].mpicc).startswith('mpi') \
+               and os.path.basename(spec['mpi'].mpicxx).startswith('mpi') \
+               and os.path.basename(spec['mpi'].mpifc).startswith('mpi'):
+                # if MPI-wrappers are used we can assume that linking works without
+                # manually specifying libs, thus, pass an empty list of libraries
                 config_args.append('--with-mpi-libs= ')
+            else:
+                # we do not use MPI-wrappers, thus, it is likely that we need to
+                # manually add libs for linking; try to find all possible ones
+                mpi_static_libs = find_libraries('lib*mpi*', spec['mpi'].prefix.lib,
```

Comment:
Don’t we have spec[mpi].libs() ? It’s a job of MPI package to know its libs, not the packages that use it.

Reply:
I have only seen mpicxx_shared_libs. And I know that @adamjstewart suggested in https://github.com/spack/spack/issues/8979#issuecomment-412992210 to extend the MPI packages. But that's not trivial at all. For example the current intel-mpi package even seems  to not correctly take into account in its mpi_libs() or mpicxx_shared_libs whether the multi-threaded versions of the library (*_mt.so) are to be used or not.
MPIs typically provide MPI-compiler-wrappers which take all these specialties into account. Thus, in the suggested patch, the 'else' branch will not be used at all for Open MPI, Intel MPI, mpich, mvapich, Sectrum MPI, ... It will only be reached on nowadays exotic systems (I'm not aware of any as e.g. even Blue Gene uses wrapper these days; maybe Cray?) which do not have compiler wrappers and require manual linking with -lmpi or whatever.
Moreover, Zoltan seems to need the Fortran (MPI) libraries to be linked with a C compiler which is not a typical use case.

Reply:
I just saw that even on Cray 'ftn', 'cc', and 'CC' seem to be MPI-wrappers ...; there is no need to specify anything for MPI linking.
If there are no machines / configurations without MPI wrappers, one probably could go completely without the handcrafted wrapper check ("starts with mpi") and omit the else part; just always doing `config_args.append('--with-mpi-libs= ')` only.